### PR TITLE
Fix #2644 Server Crash

### LINKED ISSUE #2644: Object reference not set to an instance of an object - Docker
### Version

Latest 1.4-stable

### OS

Linux

### Platform

Steam

### This bug affects

Gameplay as a Player

### Description

System.ComponentModel.Win32Exception: Permission denied

### Log File

[server.log](https://github.com/tModLoader/tModLoader/files/8998297/server.log)

### Steps to reproduce

1. Create Dockerfile like below (wip)
    ```Dockerfile
    FROM ubuntu:20.04
    
    # install ca-certificates to be able to pull https traffic to...
    # install git, unzip, wget to install vanilla server, tModLoader
    # hadolint ignore=DL3008
    RUN apt-get update && apt-get install --no-install-recommends -y \
      ca-certificates \
      curl \
      git \
      unzip \
      wget \
      && apt-get clean \
      && rm -rf /var/lib/apt/lists/*
    
    # update certs
    RUN update-ca-certificates
    
    # install dotnet, required for start-tModLoaderServer.sh
    # see https://docs.microsoft.com/en-us/dotnet/core/install/linux-ubuntu#2204
    ENV DEBIAN_FRONTEND=noninteractive
    # hadolint ignore=DL3008
    RUN wget --progress=dot:giga https://packages.microsoft.com/config/ubuntu/20.04/packages-microsoft-prod.deb -O packages-microsoft-prod.deb \
      && dpkg -i packages-microsoft-prod.deb \
      && rm packages-microsoft-prod.deb \
      && apt-get update \
      && apt-get install --no-install-recommends -y apt-transport-https \
      && apt-get update \
      && apt-get install --no-install-recommends -y dotnet-sdk-6.0 \
      && apt-get clean \
      && rm -rf /var/lib/apt/lists/*
    
    # install mono
    # see https://www.mono-project.com/download/stable/#download-lin
    SHELL ["/bin/bash", "-o", "pipefail", "-c"]
    # hadolint ignore=DL3008
    RUN apt-get update \
      && apt-get install --no-install-recommends -y gnupg ca-certificates \
      && apt-key adv --keyserver hkp://keyserver.ubuntu.com:80 --recv-keys 3FA7E0328081BFF6A14DA29AA6A19B38D3D831EF \
      && echo "deb https://download.mono-project.com/repo/ubuntu stable-focal main" | tee /etc/apt/sources.list.d/mono-official-stable.list \
      && apt-get update \
      && apt-get install  --no-install-recommends -y mono-complete \
      && apt-get clean \
      && rm -rf /var/lib/apt/lists/*
    
    # install steamcmd
    # Insert Steam prompt answers
    RUN echo steam steam/question select "I AGREE" | debconf-set-selections \
      && echo steam steam/license note '' | debconf-set-selections
    # hadolint ignore=DL3008
    RUN apt-get update \
      && apt-get install -y --no-install-recommends  software-properties-common \
      && add-apt-repository multiverse \
      && dpkg --add-architecture i386 \
      && apt-get update \
      && echo "I AGREE" | apt-get install -y --no-install-recommends lib32gcc-s1 steamcmd \
      && apt-get clean \
      && rm -rf /var/lib/apt/lists/*
    
    # non-root user
    ARG USERNAME=terraria
    ARG USER_UID=1000
    ARG USER_GID=$USER_UID
    # hadolint ignore=DL3008
    RUN groupadd --gid $USER_GID $USERNAME \
      && useradd --uid $USER_UID --gid $USER_GID -m $USERNAME \
      #
      # [Optional] Add sudo support. Omit if you don't need to install software after connecting.
      && apt-get update \
      && apt-get install --no-install-recommends -y sudo \
      && apt-get clean \
      && rm -rf /var/lib/apt/lists/* \
      && echo $USERNAME ALL=\(root\) NOPASSWD:ALL > /etc/sudoers.d/$USERNAME \
      && chmod 0440 /etc/sudoers.d/$USERNAME
    USER $USERNAME
    WORKDIR /home/${USERNAME}
    
    # see https://github.com/ldericher/tmodloader-docker/blob/e24ae0950544a27fe4dc383058218766d0ecf16f/Dockerfile
    # get vanilla server \
    RUN wget --progress=dot:giga https://terraria.org/api/download/pc-dedicated-server/terraria-server-1436.zip \
      && unzip terraria-server-*.zip \
      && rm terraria-server-*.zip \
      && cp --verbose -a 1436/. . \
      && rm -rf 1436 Mac Windows \
      && mv Linux Terraria
    WORKDIR /home/${USERNAME}/
    # clone tModLoader
    # see https://github.com/tModLoader/tModLoader/wiki/Starting-a-modded-server
    ARG TMOD_LOADER_GIT_TAG=v2022.05.103.34
    RUN wget --progress=dot:giga https://github.com/tModLoader/tModLoader/releases/download/v2022.05.103.34/tModLoader.zip \
      && unzip tModLoader.zip -d tModLoader \
      && rm tModLoader.zip \
      && chmod u+x tModLoader/start-tModLoader* \
      && chmod u+x tModLoader/LaunchUtils/ScriptCaller.sh 
    
    # follow instructions at https://github.com/tModLoader/tModLoader/wiki/Starting-a-modded-server#starting-server-headless
    
    EXPOSE 7777
    CMD ["./tModLoader/start-tModLoader.sh"]
    ```

1. Create docker-compose.yaml like below
    
    ```yaml
    version: '3.8'
    services:
      terraria:
        platform: linux/amd64
        build: .
        ports:
          - 7777:7777
        command:
          - ./tModLoader/LaunchUtils/ScriptCaller.sh 
          - -server
          - -config
          - /home/terraria/serverconfig.txt
        volumes:
          - ./serverconfig.txt:/home/terraria/serverconfig.txt:ro
          - ./worlds:/home/terraria/worlds/
          - ./logs:/home/terraria/tModLoader/tModLoader-Logs/
          - ./dotnet:/home/terraria/tModLoader/dotnet/
    ```

1. Copy serverconfig.txt below

    ```
    #this is an example config file for TerrariaServer.exe
    #use the command 'TerrariaServer.exe -config serverconfig.txt' to use this configuration or run start-server.bat
    #please report crashes by emailing crashlog.txt to support@terraria.org
    
    #the following is a list of available command line parameters:
    
    #-config <config file>                                      Specifies the configuration file to use.
    #-port <port number>                                          Specifies the port to listen on.
    #-players <number> / -maxplayers <number>       Sets the max number of players
    #-pass <password> / -password <password>        Sets the server password
    #-world <world file>                                          Load a world and automatically start the server.
    #-autocreate <#>                                                Creates a world if none is found in the path specified by -world. World size is specified by: 1(small), 2(medium), and 3(large).
    #-banlist <path>                                                Specifies the location of the banlist. Defaults to "banlist.txt" in the working directory.
    #-worldname <world name>                                Sets the name of the world when using -autocreate.
    #-secure                                                                Adds addition cheat protection to the server.
    #-noupnp                                                                Disables automatic port forwarding
    #-steam                                                                 Enables Steam Support
    #-lobby <friends> or <private>             Allows friends to join the server or sets it to private if Steam is enabled
    #-ip <ip address>       Sets the IP address for the server to listen on
    #-forcepriority <priority>      Sets the process priority for this task. If this is used the "priority" setting below will be ignored.
    #-disableannouncementbox   Disables the text announcements Announcement Box makes when pulsed from wire.
    #-announcementboxrange <number>   Sets the announcement box text messaging range in pixels, -1 for serverwide announcements.
    #-seed <seed> Specifies the world seed when using -autocreate
    
    #remove the # in front of commands to enable them.
    
    #Load a world and automatically start the server.
    world=/home/terraria/worlds/world1.wld
    
    #Creates a new world if none is found. World size is specified by: 1(small), 2(medium), and 3(large).
    autocreate=1
    
    #Sets the world seed when using autocreate
    #seed=AwesomeSeed
    
    #Sets the name of the world when using autocreate
    worldname=Terraria
    
    #Sets the difficulty of the world when using autocreate 0(normal), 1(expert)
    difficulty=1
    
    #Sets the max number of players allowed on a server.  Value must be between 1 and 255
    maxplayers=8
    
    #Set the port number
    port=7777
    
    #Set the server password
    password=assass
    
    #Set the message of the day
    #motd=Please don�t cut the purple trees!
    
    #Sets the folder where world files will be stored
    #worldpath=/home/terraria/worlds/
    
    #The location of the banlist. Defaults to "banlist.txt" in the working directory.
    #banlist=banlist.txt
    
    #Adds addition cheat protection.
    secure=1
    
    #Sets the server language from its language code. 
    #English = en-US, German = de-DE, Italian = it-IT, French = fr-FR, Spanish = es-ES, Russian = ru-RU, Chinese = zh-Hans, Portuguese = pt-BR, Polish = pl-PL,
    language=en-US
    
    #Automatically forward ports with uPNP
    #upnp=1
    
    #Reduces enemy skipping but increases bandwidth usage. The lower the number the less skipping will happen, but more data is sent. 0 is off.
    #npcstream=60
    
    #Default system priority 0:Realtime, 1:High, 2:AboveNormal, 3:Normal, 4:BelowNormal, 5:Idle
    priority=3
    ```  

  
1. Build and run with the local mounts unpopulated

    ```bash
    docker-compose up --build -d && docker-compose logs -f
    ```

### Expected Behavior

Server starts and I can connect via tModLoader at 0.0.0.0:7777

### Actual Behavior

1. Observe error in log tail

    ```bash
    terraria-terraria-1  | : Object reference not set to an instance of an object. (see server.log for full trace)
    ```

1. Observe error in `./logs/server.log`
    
    ```
    [06:26:09] [Main Thread/ERROR] [tML]: Steam Game Server failed to Init. Steam Workshop downloading on GoG is unavailable. Make sure Steam is installed
    [06:26:10] [Main Thread/WARN] [tML]: Silently Caught Exception: 
    System.ComponentModel.Win32Exception: Permission denied
       at System.Diagnostics.Process.set_PriorityClassCore(ProcessPriorityClass value)
       at System.Diagnostics.Process.set_PriorityClass(ProcessPriorityClass value)
       at Terraria.Initializers.LaunchInitializer.LoadServerParameters(Main game) in tModLoader\Terraria\Initializers\LaunchInitializer.cs:line 116
       at Terraria.Initializers.LaunchInitializer.LoadParameters(Main game) in tModLoader\Terraria\Initializers\LaunchInitializer.cs:line 17
       at Terraria.Program.LaunchGame_(Boolean isServer) in tModLoader\Terraria\Program.cs:line 226
       at Terraria.Program.LaunchGame(String[] args, Boolean monoArgs) in tModLoader\Terraria\Program.cs:line 194
       at MonoLaunch.<>c__DisplayClass2_0.<Main>b__0() in tModLoader\Terraria\MonoLaunch.cs:line 65
       at MonoLaunch.Main(String[] args) in tModLoader\Terraria\MonoLaunch.cs:line 72
    ```
    
    This is despite having installed `steamcmd` per https://github.com/tModLoader/tModLoader/blob/1.4/patches/tModLoader/Terraria/release_extras/DedicatedServerUtils/Instructions.txt#L4
  
    ```bash
    $ docker-compose exec terraria bash
    terraria@76751768096b:~$ apt list | grep steam
    
    WARNING: apt does not have a stable CLI interface. Use with caution in scripts.
    
    steamcmd/now 0~20180105-3 i386 [installed,local]
    ```
  
### Reproduction frequency

100%

### Additional Information

Tested with macOS Monterey
MacBook Pro (16-inch, 2021)
Chip Apple M1 Pro
Memory 32 GB

Docker Desktop Version 4.9.1 (81317)
Engine: 20.10.16
Compose: v2.6.0
Credential Helper: v0.6.4
Kubernetes: v1.24.0
Snyk: v1.827.0

### PR DESCRIPTION
**Pull request by Solxanich**

### CODE REVIEW

File: patches/tModLoader/Terraria/Main.cs.patch [hunk 1]
```diff
@@ -831,13 +831,14 @@
...
  			YouCanSleepNow();
-@@ -4538,8 +_,14 @@
+@@ -4538,8 +_,15 @@
  
  		public static void startDedInputCallBack() {
  			while (!Netplay.Disconnect) {
 -				Console.Write(": ");
  				string text = Console.ReadLine();
-+				ExecuteCommand(text, new ConsoleCommandCaller());
++				if (text != null) // Fix for #2644 where Docker woul steal lines from the active console.
```
Comment (ghost) [inline patches/tModLoader/Terraria/Main.cs.patch]: nitpick, but "would"*

Comment (hkiang01) [reply patches/tModLoader/Terraria/Main.cs.patch:840]: [tested, works](https://github.com/tModLoader/tModLoader/issues/2644#issuecomment-1173151388)

**Comment by hkiang01** [comment]
[tested, seems to work](https://github.com/tModLoader/tModLoader/issues/2644#issuecomment-1173151388)

**Comment by Chicken-Bones** [comment]
This could be problematic, because it turns a blocking loop into a hot-loop if `Console.ReadLine` returns `null` immediately, this could end up pinning a CPU core. If there's no console, maybe we should just log a note and exit instead?

**Comment by Chicken-Bones** [comment]
@hkiang01 Can you test #2689 please?

**Comment by hkiang01** [comment]
@Chicken-Bones doesn't seem to work...

tested similarly to [earlier](https://github.com/tModLoader/tModLoader/issues/2644#issuecomment-1173151388)

last few lines of container logs
```
terraria_1  | Dotnet should be present in "/home/terraria/tModLoader/dotnet/6.0.0"
terraria_1  | Attempting Launch...
terraria_1  | sed: can't read /proc/0/cmdline: No such file or directory
terraria_1  | Launched Using Local Dotnet
```

last few lines of server.log
```
[20:40:29] [Main Thread/INFO] [tML]: Save Are Located At: /home/terraria/.local/share/Terraria/tModLoader-dev
[20:40:29] [Main Thread/INFO] [tML]: Starting tModLoader server 1.4.3.6+9999.0|dockernoninteractivefix2|dev|5d7bca8df27b501cbade13ddaf5a5a6a0e349d6f|5249612433474718825 built 07/05/2022 18:11
[20:40:29] [Main Thread/INFO] [tML]: Log date: 07/05/2022
[20:40:29] [Main Thread/INFO] [tML]: Running on Linux (v5.10.104.0) X64 NetCore 6.0.0
[20:40:29] [Main Thread/INFO] [tML]: Executable: /home/terraria/tModLoader/tModLoader.dll
[20:40:29] [Main Thread/INFO] [tML]: Working Directory: /home/terraria/tModLoader
[20:40:29] [Main Thread/INFO] [tML]: Launch Parameters: -server -config /home/terraria/serverconfig.txt
[20:40:29] [Main Thread/INFO] [tML]: Parsed Launch Parameters: -server -config /home/terraria/serverconfig.txt
[20:40:31] [Main Thread/DEBUG] [tML]: Assembly Resolve: tModLoader, Version=1.4.3.6, Culture=neutral, PublicKeyToken=null -> FNA, Version=22.7.0.0, Culture=neutral, PublicKeyToken=null
```

relevant commits:
```zsh
$ docker-compose exec terraria cat /home/terraria/Release-Build/RecentGitHubCommits.txt
4868e674b075f6d21762d7dbebfb2d77a61c1844 Update comment in ExampleShiftClickSlotPlayer
ca6743c559cfa7675e9f0e01d5e3c98a570f0606 Fix ExampleShiftClickSlotPlayer being incompatible with other usages of the hook
f9c647062af57ea7c800c4b1461042f0c82f708a Update FNA submodule
7be76be18f263bc7d4041328e196a517ed179f86 Remove `Read/WriteVarInt` extensions (#2678)
597a2cc1b1a36613b935349602b61daf5e5a20e1 Use IAudio backend again, this time with speakermask fix (#2687)
37952d81983f1570c78573669689de30103fbfc5 Fix PlayerLoader since it wasn't caught by SafeToTest label
c86f63ddcb62260c9f181b0c7d31e0917b9925ac Add PlayerLoader.HoverSlot hook (#2620)
d456e639b24d716bc4d44e8b12d1d40555e1ff5f Add support for Linux x86 formally by adding libsteamapi.so
4e3fdda4d1afb155b19f9dd98509e9987d030087 Fix #2644 Server Crash (#2672)
c81047c72b59ad376ed76c66f706399ac63ea5c9 exclude NoSync mods which don't exist on client from the ModNetDiagnosticsUI, fixing a client exception (#2676)
```

**Comment by Chicken-Bones** [comment]
Does the server just crash with no error? What about `Natives.log`?

**Comment by hkiang01** [comment]
all of natives.log

```
terminate called after throwing an instance of 'PAL_SEHException'
qemu: uncaught target signal 6 (Aborted) - core dumped
```

**Pull request merged**